controllers/upgrade_handlers.go

```diff
@@ -312,3 +309,48 @@ func (r *ImageBasedUpgradeReconciler) handleRestore(ctx context.Context) (ctrl.R
 	r.Log.Info("OADP path removed", "path", backuprestore.OadpPath)
 	return doNotRequeue(), nil
 }
+
+func (UpgHandler) IsPrePivot(rr reconcile.Reconciler, ibu *lcav1alpha1.ImageBasedUpgrade) (bool, error) {
+	r := rr.(*ImageBasedUpgradeReconciler)
+	currentStaterootName, err := r.RPMOstreeClient.GetCurrentStaterootName()
+
+	if err != nil {
+		return false, err
+	}
+	r.Log.Info("stateroots", "current stateroot:", currentStaterootName, "desired stateroot", getDesiredStaterootName(ibu))
+	return currentStaterootName != getDesiredStaterootName(ibu), nil
+}
+
+func (UpgHandler) RebootToNewStateRoot(rr reconcile.Reconciler, ibu *lcav1alpha1.ImageBasedUpgrade) error {
+	r := rr.(*ImageBasedUpgradeReconciler)
+
+	r.Log.Info("rebooting to a new stateroot")
+	r.Recorder.Event(ibu, v1.EventTypeNormal, "Reboot", "System will now reboot")
+
+	_, err := r.Executor.Execute("systemctl", "--message=\"Image Based Upgrade\"", "reboot")
+	if err != nil {
+		return err
+	}
+
+	r.Log.Info(fmt.Sprintf("Wait for %s to be killed via SIGTERM", defaultRebootTimeout.String()))
+	time.Sleep(defaultRebootTimeout)
+
+	return fmt.Errorf("failed to reboot. This should never happen! Please check the system")
+}
+
+func (UpgHandler) ExportIBUCR(ibu *lcav1alpha1.ImageBasedUpgrade, stateRootRepo string) error {
+	filePath := filepath.Join(stateRootRepo, utils.IBUFilePath)
+	// Temporarily empty resource version so the file can be used to restore status
+	rv := ibu.ResourceVersion
```

Comment:
https://github.com/openshift-kni/lifecycle-agent/pull/189 revises some of these, including:
* moving the ResourceVersion to the restore side to avoid clearing it everywhere it's saved
* renaming isPrePivot to isOrigStaterootBooted, for clarity due to rollback usage